### PR TITLE
Add windows guest support for upload/download data via ipv6

### PR DESCRIPTION
**Pull request by pezhang**
1. Add get windows ipv6 locallink address
2. Enable upload/download big data between host and windows guest

ID: 1566963
Signed-off-by: Pei Zhang <pezhang@redhat.com>

### CODE REVIEW

**Comment by pezhang** [comment]
Depend on:
https://github.com/avocado-framework/avocado-vt/pull/1588

**Comment by pezhang** [comment]
(1/1) Host_RHEL.m7.u6.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.bigfile_transfer_over_ipv6: PASS (1538.78 s)

 (1/1) Host_RHEL.m7.u6.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.bigfile_transfer_over_ipv6: PASS (1433.00 s)

File: qemu/tests/transfer_file_over_ipv6.py [hunk 1]
```diff
@@ -26,25 +29,25 @@ def run(test, params, env):
     port = params.get("file_transfer_port")
     password = params.get("password")
     username = params.get("username")
-    tmp_dir = params.get("tmp_dir", "/tmp/")
+    if params.get("os_type") == "linux":
+        tmp_dir = params.get("tmp_dir", "/tmp/")
```
Comment (xiywang) [inline qemu/tests/transfer_file_over_ipv6.py]: suggest to put 'tmp_dir' in cfg file and set for different os type.

File: qemu/tests/transfer_file_over_ipv6.py [hunk 2]
```diff
@@ -72,14 +75,28 @@ def get_file_md5sum(file_name, session, timeout):
...
+            inet_name[vm] = utils_net.get_linux_ifname(
+                sessions[vm], vm.get_mac_address())
+            addresses[vm] = utils_net.get_guest_ip_addr(
+                sessions[vm],
+                vm.get_mac_address(),
+                params.get("os_type"),
+                ip_version="ipv6",
+                linklocal=True)
+        else:
+            addresses[vm] = get_windows_ipv6_linklocal_address(sessions[vm])
```
Comment (xiywang) [inline qemu/tests/transfer_file_over_ipv6.py]: any reason not to use utils_net.get_guest_ip_addr to get guest ip in win guest?

File: qemu/tests/transfer_file_over_ipv6.py [hunk 3]
```diff
@@ -100,42 +119,50 @@ def get_file_md5sum(file_name, session, timeout):
...
+                        remote.scp_between_remotes(addresses[vm_src],
+                                                   addresses[vm_dst],
+                                                   port, password, password,
+                                                   username, username,
+                                                   guest_path, dest_path,
+                                                   timeout=file_trans_timeout,
+                                                   src_inter=host_ifname,
+                                                   dst_inter=inet_name[vm_src])
+                        dst_md5 = get_file_md5sum(
+                            dest_path, sessions[vm_dst], timeout=file_md5_check_timeout)
```
Comment (xiywang) [inline qemu/tests/transfer_file_over_ipv6.py]: pay attention to the indent. 
the original one is acceptable.

File: qemu/tests/transfer_file_over_ipv6.py [hunk 4]
```diff
@@ -72,14 +75,28 @@ def get_file_md5sum(file_name, session, timeout):
...
+                          logging.info)
     for vm in vms:
         vm.verify_alive()
         sessions[vm] = vm.wait_for_login(timeout=timeout)
-        inet_name[vm] = utils_net.get_linux_ifname(sessions[vm],
-                                                   vm.get_mac_address())
-        addresses[vm] = get_linux_ipv6_linklocal_address(inet_name[vm],
-                                                         sessions[vm])
+        if params.get("os_type") == "linux":
+            inet_name[vm] = utils_net.get_linux_ifname(
```
Comment (xiywang) [inline qemu/tests/transfer_file_over_ipv6.py]: indent

File: qemu/tests/transfer_file_over_ipv6.py [hunk 5]
```diff
@@ -72,14 +75,28 @@ def get_file_md5sum(file_name, session, timeout):
 
     # config ipv6 address host and guest.
     host_ifname = params.get("netdst")
+    host_address = utils_net.get_host_ip_address(
```
Comment (xiywang) [inline qemu/tests/transfer_file_over_ipv6.py]: could you explain why need to add the param "linklocal"?

File: qemu/tests/transfer_file_over_ipv6.py (hunk 1 above)
Comment (pezhang) [reply qemu/tests/transfer_file_over_ipv6.py]: @xiywang Thanks for review. Updated.

File: qemu/tests/transfer_file_over_ipv6.py (hunk 2 above)
Comment (pezhang) [reply qemu/tests/transfer_file_over_ipv6.py]: @xiywang  Changed avocado-vt to support windows guest for ipv6 locallink. So already updated to use utils_net.get_guest_ip_addr for win guest.

File: qemu/tests/transfer_file_over_ipv6.py (hunk 3 above)
Comment (pezhang) [reply qemu/tests/transfer_file_over_ipv6.py]: @xiywang Updated.

File: qemu/tests/transfer_file_over_ipv6.py (hunk 5 above)
Comment (pezhang) [reply qemu/tests/transfer_file_over_ipv6.py]: @xiywang Updated.

File: qemu/tests/transfer_file_over_ipv6.py (hunk 4 above)
Comment (pezhang) [reply qemu/tests/transfer_file_over_ipv6.py]: @xiywang Updated.

**Comment by pezhang** [comment]
@xiywang @sitoliu Please help to review again. Thanks.

File: qemu/tests/transfer_file_over_ipv6.py [hunk 6]
```diff
@@ -26,29 +29,12 @@ def run(test, params, env):
     port = params.get("file_transfer_port")
     password = params.get("password")
     username = params.get("username")
-    tmp_dir = params.get("tmp_dir", "/tmp/")
+    tmp_dir = params.get("tmp_dir")
```
Comment (xiywang) [inline qemu/tests/transfer_file_over_ipv6.py]: better to give a default value, or if you're sure that this parameter definitely have been set in cfg file, you can use params["tmp_dir"]

File: qemu/tests/transfer_file_over_ipv6.py [hunk 7]
```diff
@@ -100,42 +100,50 @@ def get_file_md5sum(file_name, session, timeout):
                                  host_path, guest_path,
                                  timeout=file_trans_timeout,
                                  interface=host_ifname)
-            dst_md5 = get_file_md5sum(guest_path, sessions[vm],
-                                      timeout=file_md5_check_timeout)
+            dst_md5 = get_file_md5sum(
+                guest_path, sessions[vm], timeout=file_md5_check_timeout)
```
Comment (xiywang) [inline qemu/tests/transfer_file_over_ipv6.py]: typo. 
params should align with "(".
suggest to use dst_md5 = get_file_md5sum(guest_path, sessions[vm],
                                                                       timeout=file_md5_check_timeout) (align with the "(" in above line)

File: qemu/tests/transfer_file_over_ipv6.py (hunk 6 above)
Comment (pezhang) [reply qemu/tests/transfer_file_over_ipv6.py]: This option has been definitely set in it's cfg file. Thanks for review.

Comment (xiywang) [reply qemu/tests/transfer_file_over_ipv6.py]: use params["tmp_dir"] instead of .get()

**Comment by pezhang** [comment]
@xiywang , updated. Please help to review again. Thanks.

File: qemu/tests/cfg/transfer_file_over_ipv6.cfg [hunk 8]
```diff
@@ -1,10 +1,13 @@
...
     type = transfer_file_over_ipv6
-    vms += " vm2"
+    vms = "vm1 vm2"
     image_snapshot = yes
     filesize = 4096
     file_trans_timeout = 2400
     file_md5_check_timeout = 600
     dd_cmd = "dd if=/dev/zero of=%s oflag=direct bs=1M count=%d"
+    Linux:
+        tmp_dir = "/tmp/"
```
Comment (sitoliu) [inline qemu/tests/cfg/transfer_file_over_ipv6.cfg]: Confirm with @yanan-fu , RHEL 8.0 guest make `/tmp` as a seperate partition, '3.4G' is the total size now, please use `/var/tmp` as your temp dir. Thanks.

**Comment by pezhang** [comment]
@sitoliu updated. Please help to review again. Thanks.

**Comment by pezhang** [comment]
In rhel8, there might be a product bug[1], that's why Sitong's above testing fail.

[1]Bug 1620333 - scp between two remote servers fail with ipv6

**Comment by vivianQizhu** [comment]
@pezhang @sitoliu Would anyone provide latest test result for both Windows and Linux guest? Thanks.

**Comment by pezhang** [comment]
@vivianQizhu Testing it now. Will update the testing results once finish. 

Thank you.

**Comment by pezhang** [comment]
(4/5) Host_RHEL.m8.u0.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.bigfile_transfer_over_ipv6: PASS (737.83 s)
 (5/5) Host_RHEL.m8.u0.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.bigfile_transfer_over_ipv6: PASS (809.88 s)

**Pull request merged**